### PR TITLE
#184347090-two-factor-authentication-for-sellers

### PR DESCRIPTION
**Pull request by Ntarecedrick**
### What does this PR do?

this PR is to add the two factor authentication to our e-commerce website

### Description of Task to be completed?

the task was to send a two factor authentication for sellers who are login in our website 

### How should this be manually tested?

you need to login and assign yourself a role of a seller and then once you are assigned as a seller upon successful login a seller is to be sent a TOTP message on his/her email. 

### Any background context you want to provide?

N/A 

### What are the relevant pivotal tracker stories?

[Provide a relevant link to the pivotal tracker.](https://www.pivotaltracker.com/n/projects/2628461/stories/184347090)

### Screenshots (if appropriate)

N/A

### Questions:

# Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or Enhancement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] My code is formatted correctly
- [x] My changes doesn't have any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by NG-Solennel — DISMISSED** [review]

**Review by teerenzo — DISMISSED** [review]

**Review by Marthatwesi — DISMISSED** [review]

File: src/controllers/2FAcontroller.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+/* eslint-disable no-else-return */
```
Comment (Eli250) [inline src/controllers/2FAcontroller.js:1]: Put these contents in a user controller file

File: src/database/migrations/20230131072323-create-user.js [hunk 2]
```diff
@@ -1,33 +0,0 @@
-/** @type {import('sequelize-cli').Migration} */
```
Comment (Eli250) [inline src/database/migrations/20230131072323-create-user.js:1]: Why did you delete this file?

File: src/routes/api/2FA.route.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import express from "express";
```
Comment (Eli250) [inline src/routes/api/2FA.route.js:1]: Move these into user route

File: src/routes/index.js [hunk 4]
```diff
@@ -1,9 +1,11 @@
 import express, { Router } from "express";
+import twoFactorAuthentication from "./api/2FA.route";
 import home from "./api/home.route";
 import user from "./api/user.route";
 
 const routes = express.Router();
 
 routes.use("/home", home);
 routes.use("/", user);
+routes.use("/", twoFactorAuthentication);
```
Comment (Eli250) [inline src/routes/index.js:10]: Add this to user route

**Review by Eli250 — CHANGES_REQUESTED** [review]

File: src/controllers/2FAcontroller.js (hunk 1 above)
Comment (Ntarecedrick) [reply src/controllers/2FAcontroller.js:1]: > Put these contents in a user controller file

the whole content of the 2FA controller ? i thought the two factor controller has its own controller

File: src/routes/index.js (hunk 4 above)
Comment (Ntarecedrick) [reply src/routes/index.js:10]: > Add this to user route

isn't the two factor authenticator an independent route tho ?

File: src/database/migrations/20230131072323-create-user.js (hunk 2 above)
Comment (Ntarecedrick) [reply src/database/migrations/20230131072323-create-user.js:1]: > Why did you delete this file?

i don't think i did it on purpose but let me check

**Review by Eli250 — CHANGES_REQUESTED** [review]

File: src/services/2FA.service.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  static async checkUser(data) {
+    const email = data;
+    const user = await User.findOne({ where: { email } });
+    if (user.role === "seller" || user.role === "Admin") {
+      const { name } = user;
+      const randomAuth = generateRandom();
+      const Options = {
+        expiresIn: "45000",
+      };
+      const token = JwtUtil.generate({ name, email, randomAuth }, Options);
```
[the remark this answers is not in the document]
Comment (Ntarecedrick) [reply src/services/2FA.service.js:18]: > You have to do this in a controller

okay let me work on it

**Review by NG-Solennel — DISMISSED** [review]

**Review by Eli250 — APPROVED** [review]
